File: src/main/java/at/hrechny/predictionsbot/database/repository/SeasonRepository.java
package at.hrechny.predictionsbot.database.repository;

import at.hrechny.predictionsbot.database.entity.SeasonEntity;
import jakarta.inject.Singleton;
import jakarta.persistence.EntityManager;
import jakarta.transaction.Transactional;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Singleton
public class SeasonRepository {

  private final EntityManager entityManager;

  public SeasonRepository(EntityManager entityManager) {
    this.entityManager = entityManager;
  }

  @Transactional
  public SeasonEntity save(SeasonEntity entity) {
    return entityManager.merge(entity);
  }

  public Optional<SeasonEntity> findById(UUID id) {
    return Optional.ofNullable(entityManager.find(SeasonEntity.class, id));
  }

  public List<SeasonEntity> findAllByActiveIsTrue() {
    return entityManager.createQuery("select s from SeasonEntity s where s.active = true", SeasonEntity.class).getResultList();
  }

  public List<SeasonEntity> findAllByCompetitionId(UUID competitionId) {
    return entityManager
        .createQuery("select s from SeasonEntity s where s.competition.id = :competitionId", SeasonEntity.class)
        .setParameter("competitionId", competitionId)
        .getResultList();
  }

  public Optional<SeasonEntity> findFirstByCompetitionIdAndActiveIsTrue(UUID competitionId) {
    return entityManager
        .createQuery("select s from SeasonEntity s where s.competition.id = :competitionId and s.active = true", SeasonEntity.class)
        .setParameter("competitionId", competitionId)
        .setMaxResults(1)
        .getResultStream()
        .findFirst();
  }

  public int countAllByActiveIsTrueAndCompetitionId(UUID competitionId) {
    return entityManager
        .createQuery("select count(s) from SeasonEntity s where s.active = true and s.competition.id = :competitionId", Long.class)
        .setParameter("competitionId", competitionId)
        .getSingleResult()
        .intValue();
  }
}

File: src/main/kotlin/at/hrechny/predictionsbot/database/repository/SeasonRepository.kt
package at.hrechny.predictionsbot.database.repository

import at.hrechny.predictionsbot.database.entity.SeasonEntity
import jakarta.inject.Singleton
import jakarta.persistence.EntityManager
import io.micronaut.transaction.annotation.Transactional
import java.util.Optional
import java.util.UUID

@Singleton
@Transactional
class SeasonRepository(
    private val entityManager: EntityManager,
) {
    @Transactional
    fun save(entity: SeasonEntity): SeasonEntity = entityManager.merge(entity)

    fun findById(id: UUID): Optional<SeasonEntity> = Optional.ofNullable(entityManager.find(SeasonEntity::class.java, id))

    fun findAllByActiveIsTrue(): List<SeasonEntity> =
        entityManager.createQuery("select s from SeasonEntity s where s.active = true", SeasonEntity::class.java).resultList

    fun findAllByCompetitionId(competitionId: UUID): List<SeasonEntity> = entityManager
        .createQuery("select s from SeasonEntity s where s.competition.id = :competitionId", SeasonEntity::class.java)
        .setParameter("competitionId", competitionId)
        .resultList

    fun findFirstByCompetitionIdAndActiveIsTrue(competitionId: UUID): Optional<SeasonEntity> = entityManager
        .createQuery(
            "select s from SeasonEntity s where s.competition.id = :competitionId and s.active = true",
            SeasonEntity::class.java,
        )
        .setParameter("competitionId", competitionId)
        .setMaxResults(1)
        .resultStream
        .findFirst()

    fun countAllByActiveIsTrueAndCompetitionId(competitionId: UUID): Int = entityManager
        .createQuery(
            "select count(s) from SeasonEntity s where s.active = true and s.competition.id = :competitionId",
            Long::class.javaObjectType,
        )
        .setParameter("competitionId", competitionId)
        .singleResult
        .toInt()
}
